src/bundle.c

```diff
@@ -575,34 +616,83 @@ enum swupd_code remove_bundles(struct list *bundles)
 				error("\nBundle \"%s\" is required by %d bundle%s, skipping it...\n", bundle, number_of_reqd, number_of_reqd == 1 ? "" : "s");
 				info("Use \"swupd bundle-remove --force %s\" to remove \"%s\" and all bundles that require it\n", bundle, bundle);
 				list_free_list_and_data(reqd_by, free);
-				reqd_by = NULL;
 				ret_code = SWUPD_REQUIRED_BUNDLE_ERROR;
 				bad++;
 				continue;
 			} else {
 				/* the --force option was specified */
-				info("\nThe --force option was used, bundle \"%s\" and all bundles that require it will be removed from the system\n", bundle);
 				char *dep;
 				for (iter = list_head(reqd_by); iter; iter = iter->next) {
 					dep = iter->data;
 					list_move_item(dep, &current_mom->submanifests, &bundles_to_remove, find_manifest);
 					remove_tracked(dep);
 				}
 				list_free_list_and_data(reqd_by, free);
-				reqd_by = NULL;
 			}
 		}
 
-		/* move the manifest of the bundle to be removed from the list of subscribed bundles
-		 * to the list of bundles to be removed */
+		/* if we have reached this point the bundle is "deletable" */
+
+		/* move the manifest of the bundle to be removed
+		 * from the list of subscribed bundles to the
+		 * list of bundles to be removed */
 		list_move_item(bundle, &current_mom->submanifests, &bundles_to_remove, find_manifest);
-		info("\nRemoving bundle: %s\n", bundle);
 		remove_tracked(bundle);
+
+		/* if recursive, the dependencies need to be removed too */
+		if (cmdline_option_recursive) {
+
+			struct list *dependencies = NULL;
+			struct list *deps = NULL;
+
+			/* get the bundle's dependencies */
+			ret = recurse_dependencies(current_mom, bundle, &dependencies, NULL);
+			if (ret) {
+				error("Cannot get dependencies for bundle \"%s\"\n", bundle);
+				list_free_list_and_data(dependencies, manifest_free_data);
+				ret_code = ret;
+			}
+
+			/* sanitize the dependencies and add the user
+			 * provided bundles to the list so they are
+			 * considered when validating the bundle */
+			for (iter = dependencies; iter; iter = iter->next) {
+				struct manifest *m = iter->data;
+				if (is_deletable_dependency(m->component)) {
+					deps = list_prepend_data(deps, m->component);
+				}
+			}
+			deps = list_concat(deps, list_clone(bundles));
```

Comment:
Just a hit. Usually using list_clone() isn't a good idea. You can always work in the original list if it's a read-only operation. In this case all elements from bundles should be covered by recurse_manifest(), right?

Reply:
the updated Pr doesn't use list_clone anymore, thanks for the tip.